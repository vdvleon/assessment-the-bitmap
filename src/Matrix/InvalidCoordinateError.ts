
export default class InvalidCoordinateError extends Error {
    constructor(x: number, y: number, width: number, height: number) {
        super(`Invalid coordinate given: (${x}, ${y}) for matrix of size ${width} x ${height}`);
    }
}
