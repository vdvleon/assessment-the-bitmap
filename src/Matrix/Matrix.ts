import InvalidCoordinateError from "./InvalidCoordinateError";

export default class Matrix<T> {
    private width: number;
    private height: number;
    private data: T[][];

    /**
     * Initialize a matrix for a given size.
     *
     * @param width
     * @param height
     * @param initialValue
     */
    public constructor(width: number, height: number, initialValue: T) {
        this.width = width;
        this.height = height;
        this.data = [];

        // Initialize the 'matrix'
        for (let h = 0; h < height; ++h) {
            this.data[h] = [];
            for (let w = 0; w < width; ++w) {
                this.data[h][w] = initialValue;
            }
        }
    }

    public getWidth(): number {
        return this.width;
    }

    public getHeight(): number {
        return this.height;
    }

    public getData(): T[][] {
        return this.data;
    }

    /**
     * Get the complete row data (an array of T values).
     *
     * @param y Row index
     * @return Values for all columns in given row
     */
    public getRowData(y: number): T[] {
        return this.data[y];
    }

    /**
     * Set value at specific coordinate.
     *
     * @param x
     * @param y
     * @param value
     */
    public setValueAt(x: number, y: number, value: T): this {
        this.validateCoordinate(x, y);

        this.data[y][x] = value;

        return this;
    }

    /**
     * Get value at specific coordinate.
     *
     * @param x
     * @param y
     */
    public getValueAt(x: number, y: number): T {
        this.validateCoordinate(x, y);

        return this.data[y][x];
    }

    /**
     * Validate a coordinate to the size of the matrix.
     *
     * @param x
     * @param y
     */
    private validateCoordinate(x, y) {
        if (x < 0 || x >= this.width || y < 0 || y >= this.height) {
            throw new InvalidCoordinateError(x, y, this.width, this.height);
        }
    }
}
