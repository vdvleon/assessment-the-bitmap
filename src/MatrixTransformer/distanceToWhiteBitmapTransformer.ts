import Matrix from "../Matrix/Matrix";
import { MatrixTransformer } from "./MatrixTransformer";

export interface ICoordinate {
    x: number;
    y: number;
}

export type DistanceCalculator = (p1: ICoordinate, p2: ICoordinate) => number;

/**
 * Create a bitmap transformer that says for every 'coordinate' how far away it
 * is from an other 'coordinate' that is 'white' (true).
 *
 * @param distanceCalculator A function that calculates the distance between two coordinates.
 * @return MatrixTransformer from a boolean (bitmap) matrix to a number matrix
 */
export function createFromDistanceCalculator(
    distanceCalculator: DistanceCalculator,
): MatrixTransformer<boolean, number> {
    return (matrix: Matrix<boolean>): Matrix<number> => {
        // Find 'white' coordinates
        const whites = findCoordinatesForValue(matrix, true);

        // Create distance matrix
        const distanceBitmap = new Matrix<number>(
            matrix.getWidth(),
            matrix.getHeight(),
            0,
        );
        for (let y = 0; y < matrix.getHeight(); ++y) {
            for (let x = 0; x < matrix.getWidth(); ++x) {
                const distance = findClosestDistanceInList(
                    { x, y },
                    whites,
                    distanceCalculator,
                );
                distanceBitmap.setValueAt(x, y, distance);
            }
        }

        return distanceBitmap;
    };
}

/**
 * Helper to find closest distance of a coordinate to a coordinate in list.
 *
 * @param coordinate
 * @param coordinates
 * @param distance
 */
function findClosestDistanceInList(
    coordinate: ICoordinate,
    coordinates: ICoordinate[],
    distance: DistanceCalculator,
): number {
    return coordinates.reduce(
        (closestDistance, cor) =>
            Math.min(distance(coordinate, cor), closestDistance),
        +Infinity,
    );
}

/**
 * Find list of coordinates of which the value in the matrix matches a specific value.
 *
 * @param matrix
 * @param value
 * @return Coordinates
 */
function findCoordinatesForValue<T>(matrix: Matrix<T>, value: T): ICoordinate[] {
    const coordinates = new Array<ICoordinate>();

    for (let y = 0; y < matrix.getHeight(); ++y) {
        for (let x = 0; x < matrix.getWidth(); ++x) {
            if (matrix.getValueAt(x, y) === value) {
                coordinates.push({ x, y });
            }
        }
    }

    return coordinates;
}
