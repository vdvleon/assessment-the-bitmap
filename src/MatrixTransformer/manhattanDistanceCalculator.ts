import { ICoordinate } from "./distanceToWhiteBitmapTransformer";

export default function manhattenDistanceCalculator(p1: ICoordinate, p2: ICoordinate): number {
    return Math.abs(p1.x - p2.x) + Math.abs(p1.y - p2.y);
}
