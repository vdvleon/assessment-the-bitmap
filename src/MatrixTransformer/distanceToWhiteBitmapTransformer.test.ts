import Matrix from "../Matrix/Matrix";
import { createFromDistanceCalculator } from "./distanceToWhiteBitmapTransformer";
import manhattanDistanceCalculator from "./manhattanDistanceCalculator";

describe("distanceToWhiteBitmapTransformer", () => {
    describe("createFromDistanceCalculator", () => {
        it("should return a bitmap to number matrix transformer", () => {
            const transformer = createFromDistanceCalculator(manhattanDistanceCalculator);
            const input = new Matrix<boolean>(3, 5, false);
            input.setValueAt(0, 0, true);
            input.setValueAt(2, 4, true);

            const output = transformer(input);

            expect(output.getWidth()).toBe(3);
            expect(output.getHeight()).toBe(5);

            expect(output.getData()).toMatchSnapshot();
        });
    });
});
