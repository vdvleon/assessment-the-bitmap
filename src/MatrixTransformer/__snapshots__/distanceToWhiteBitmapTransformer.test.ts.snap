// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`distanceToWhiteBitmapTransformer createFromDistanceCalculator should return a bitmap to number matrix transformer 1`] = `
Array [
  Array [
    0,
    1,
    2,
  ],
  Array [
    1,
    2,
    3,
  ],
  Array [
    2,
    3,
    2,
  ],
  Array [
    3,
    2,
    1,
  ],
  Array [
    2,
    1,
    0,
  ],
]
`;
